cloudfoundry/resource_cf_service_instance.go

```diff
@@ -157,26 +168,70 @@ func resourceServiceInstanceUpdate(d *schema.ResourceData, meta interface{}) (er
 		tags = append(tags, v.(string))
 	}
 
-	_, err = sm.UpdateServiceInstance(id, name, servicePlan, params, tags)
-	return err
+	if _, err = sm.UpdateServiceInstance(id, name, servicePlan, params, tags); err != nil {
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// Check whetever service_instance exists and is in state 'succeeded'
+	if err = sm.WaitServiceInstanceTo("update", id); err != nil {
+		return err
+	}
+
+	return nil
 }
 
 func resourceServiceInstanceDelete(d *schema.ResourceData, meta interface{}) (err error) {
 
 	session := meta.(*cfapi.Session)
+	id := d.Id()
+
 	if session == nil {
 		return fmt.Errorf("client is nil")
 	}
 	session.Log.DebugMessage("begin resourceServiceInstanceDelete")
 
 	sm := session.ServiceManager()
 
-	err = sm.DeleteServiceInstance(d.Id())
-	if err != nil {
+	if err = sm.DeleteServiceInstance(id); err != nil {
+		return err
+	}
+
+	// Check whether service_instance has been indeed deleted (sometimes takes very long))
+	if err = sm.WaitDeletionServiceInstance(id); err != nil {
```

Comment:
Althrough the [CC API doc for service instance deletion](http://apidocs.cloudfoundry.org/1.40.0/service_instances/delete_a_service_instance.html) is'nt very explicit, the async deletion should return a 202 whereas a sync deletion should a 204, see [associated CC specs](https://github.com/cloudfoundry/cloud_controller_ng/blob/6fac1e18f93599c24ec471ce4d7ec4f9c1d95e9c/spec/unit/controllers/services/service_instances_controller_spec.rb#L2421-L2433)

In the case of a 204 return status, the service status polling is unnecessary

Reply:
This is a very good point. Unfortunately the status code is shallowed here:

https://github.com/mevansam/terraform-provider-cf/blob/3af2dbfd6f7f2477ebba5c1c6d8082bf2a252991/vendor/code.cloudfoundry.org/cli/cf/net/gateway.go#L178

But I come up with an alternative solution. First I try to delete the service instance via synchronous service operation, if that does not work, then asynchronous + pulling. See this commit for more details: https://github.com/mevansam/terraform-provider-cf/pull/52/commits/86a28396fea8e09d44af958b163a15b7389a2982

Reply:
I had to revert this change (https://github.com/mevansam/terraform-provider-cf/pull/52/commits/5f90fbfa5abdd2f2e92862b58c0739059a7c01f4). Unfortunately it broke  the test async provisioning, and sync deprovisioning. I tried to call this endpoint https://github.com/mevansam/terraform-provider-cf/blob/master/vendor/code.cloudfoundry.org/cli/cf/net/gateway.go#L127-L129 by rewriting this part https://github.com/chrismathias/terraform-provider-cf/blob/72c9af8aa6d6e59fb8def1c342470e21e54d874c/cloudfoundry/cfapi/service_manager.go#L555-L557 and trying the _accepts_incomplete=false_ and  _async=false_ flags, but it did not help. (see the flags here: https://apidocs.cloudfoundry.org/1.40.0/service_instances/delete_a_service_instance.html )